File: leetcode/LC/DP/0064_minimum_path_sum.py
from typing import List

#  https://leetcode.com/problems/minimum-path-sum/

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        row = len(grid)
        col = len(grid[0])
        dp = [0 for i in range(col+1)]
        for r in range(row):
            for c in range(col):
                # if it is first row just add previous dp value
                if not r:
                    dp[c + 1] = grid[r][c] + dp[c]
                # if it is first column just add current grid value
                elif not c:
                    dp[c + 1] += grid[r][c]
                # take min of value at one row less and one col less, + current grid value
                else:
                    dp[c+1] = min(dp[c+1], dp[c]) + grid[r][c]
        return dp[col]

File: leetcode/LC1/DP/0064_minimum_path_sum.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int col = grid[0].size();
        int row = grid.size();
        
        vector<int> dp (col, 0);
        dp[0] = grid[0][0];
        for(int i = 1; i < col; i++)
            dp[i] = grid[0][i] + dp[i-1];
        
        for(int i = 1; i< row; i++){
            for(int j = 0; j< col; j++){
                if(j == 0)
                    dp[j] += grid[i][j];
                else
                    dp[j] = min(dp[j], dp[j-1]) + grid[i][j];
            }
        }
        return dp[col-1];
    }
};
